### PR TITLE
Feature urn management support

### LINKED ISSUE #79: Bug in description of tags for the Falaise urn_db_service
Directory "resources/urn/db" contains a set of files that describe the dependencies between various
setup published in Falaise for simulation/reconstruction processing.
A graph analysis performed on the current description revealed some cycles in the relationships between some urn nodes related to the management and accounting of variant profiles. 
This must be fixed. The current model to manage variant profile through their URN identifier is broken.
Work to do:
- fix some of the "xxx_infos.defs" files to remove some invalid links (backward edges)
- set a valid scheme to describe correctly the dependencies between setups, variant services and variant profiles and refactor the broken links as forward edges/links
- modify the flsimulate(-configure) code to take into account these changes while searching default urn associated to a variant profile
- update the README file to describe the new rules.

A fix already exists as one of my a feature branches. Will PR it ASAP.
To be assigned to fmauger.

### PR DESCRIPTION
**Pull request by fmauger**
This work proposes a fix to issue #79.

Changes proposed in this pull request:
-  fix some of the "resources/urn/db/xxx_infos.defs" files to remove some invalid links (backward edges causing cycles)
- use a valid scheme to describe correctly the dependencies between setups, variant services and variant profiles and refactor the broken links as forward edges/links
- modify the flsimulate(-configure) code to take into account these changes while searching default urn associated to a variant profile
- update the README file to describe the new rules.
- add a preliminary version of the fltags application to generate an URN dependency graph of tagged setups.

Does this pull request fix any reported issues?
- Fixes #79 .

Additional comments or information
----------------------------------
- More documentation will be added in a future PR.
- Make ylemiere the reviewer for validation of the fix at CCIN2P3.
- Next primaries/simulation setups (#75) in preparation will be released in this new framework

### CODE REVIEW

**Comment by fmauger** [comment]
Important: Yves, review this PR before PR #75. Thanks!

**Comment by lemiere** [comment]
This fix is not compatible with Bayeux 3.0.0 due to bayeux/datatools/dependency_graph.h

build error : 
`Falaise/source/fltags/fltagsmain.cc:43:47: fatal error: bayeux/datatools/dependency_graph.h: No such file or directory`

**Comment by drbenmorgan** [comment]
As @lemiere reports, the implementation as it stands looks like it requires an updated version of Bayeux. That's fine, and we'll need to know the appropriate tag, plus a full list of changes, especially any that affect serialisation and physics outputs.

**Comment by lemiere** [comment]
I installed Bayeux3.1.1 from  https://github.com/BxCppDev/Bayeux/releases/tag/3.1.1
Root installed is : root6/6.12.04

I tried to install Falaise PR#82
But after a ninja -j8, I obtained : 

[325/383] Generating event_browser_dict.cxx, ../.....ib/Falaise/modules/libFalaise_EventBrowser.rootmap
FAILED: modules/EventBrowser/source/event_browser_dict.cxx BuildProducts/lib/Falaise/modules/libFalaise_EventBrowser_rdict.pcm BuildProducts/lib/Falaise/modules/libFalaise_EventBrowser.rootmap 
cd /home/lemiere/sofware/NEMO/simulation/3.0/Falaise_test/build/modules/EventBrowser/source && /home/lemiere/sofware/NEMO/simulation/3.0/cadfaelbrew/Cellar/cmake/3.10.2/bin/cmake -E env LD_LIBRARY_PATH=/home/lemiere/sofware/NEMO/simulation/3.0/cadfaelbrew/lib/root: /home/lemiere/sofware/NEMO/simulation/3.0/cadfaelbrew/bin/rootcling -v2 -f event_browser_dict.cxx -s /home/lemiere/sofware/NEMO/simulation/3.0/Falaise_test/build/BuildProducts/lib/Falaise/modules/libFalaise_EventBrowser.so -rml libFalaise_EventBrowser.so -rmf /home/lemiere/sofware/NEMO/simulation/3.0/Falaise_test/build/BuildProducts/lib/Falaise/modules/libFalaise_EventBrowser.rootmap -noIncludePaths -I/home/lemiere/sofware/NEMO/simulation/3.0/Falaise_test/source/Falaise/modules/EventBrowser/source/inc inc/falaise/snemo/view/event_browser_signals.h inc/falaise/snemo/view/view_models.h inc/falaise/snemo/view/event_browser.h inc/falaise/snemo/view/display_3d.h inc/falaise/snemo/view/display_2d.h inc/falaise/snemo/view/display_options.h inc/falaise/snemo/view/browser_tracks.h inc/falaise/snemo/view/event_selection.h inc/falaise/snemo/view/progress_bar.h inc/falaise/snemo/view/status_bar.h /home/lemiere/sofware/NEMO/simulation/3.0/Falaise_test/source/Falaise/modules/EventBrowser/source/inc/falaise/snemo/view/event_browser_linkdef.h
inc/falaise/snemo/view/status_bar.h /home/lemiere/sofware/NEMO/simulation/3.0/Falaise_test/source/Falaise/modules/EventBrowser/source/inc/falaise/snemo/view/event_browser_linkdef.h
input_line_12:6:10: fatal error: 'inc/falaise/snemo/view/event_browser_signals.h' file not found
#include "inc/falaise/snemo/view/event_browser_signals.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error: /home/lemiere/sofware/NEMO/simulation/3.0/cadfaelbrew/bin/rootcling: compilation failure (/home/lemiere/sofware/NEMO/simulation/3.0/Falaise_test/build/BuildProducts/lib/Falaise/modules/libFalaise_EventBrowserdc878cd013_dictUmbrella.h)

I've the file : /home/lemiere/sofware/NEMO/simulation/3.0/Falaise_test/source/Falaise/modules/EventBrowser/source/inc/falaise/snemo/view/event_browser_signals.h

**Comment by drbenmorgan** [comment]
@lemiere, that should have been fixed in #76, so this PR should be merged/rebased on `develop` to pick that update up.

**Comment by lemiere** [comment]
I successfully installed Falaise PR#82 
Updated requirements are  : 
clang
bayeux 3.1.1 
Need to merge on develop to get Update include_directories to point ROOT to our dictionary headers (self-accepted by @drbenmorgan )

Still problematic (@fmauger ):
`
/flvisualize --input-file ~/sofware/reconstruction_test_PR82.brio 
Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for inc/brio/detail/brio_record.h
   requested to autoload type TArrayCMod
Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for inc/brio/detail/brio_record.h
   requested to autoload type brio_record
terminate called after throwing an instance of 'std::logic_error'
  what():  [static void FLVisualize::FLVisualizeArgs::do_postprocess(FLVisualize::FLVisualizeArgs&):271: Cannot query URN='urn:snemo:demonstrator:geometry:4.0'!]
Aborted (core dumped)
`

File: source/fltags/FLTagsCommandLine.cc [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+      ("version", "print version number")
+
+      ("verbosity,V",
+       bpo::value<std::string>(&verbosityLabel)
+       ->value_name("level"),
+       "set the verbosity level\n"
+       "Example: \n"
+       "  -V \"debug\" ")
+
+      ("with-vertex-index,i",
```
[the remark this answers is not in the document]
Comment (fmauger) [reply source/fltags/FLTagsCommandLine.cc]: * No urn/tag names have not been changed. There is just a better/fixed description of 
  the dependencies between existing tags. The former set of tags are thus still usable.
  The tag system must be able to use any old tagged setup  during the full life cycle of 
  the Falaise software.
* A very basic documentation is added (at least a skeleton of it) about the management of resource files and the management of tags in Falaise.
* The default behaviour of fltags has been changed.

**Comment by lemiere** [comment]
Hi,

I installed  Falaise PR#82 based on  : 
Bayeux 3.1.1
Clang
xdot
We need to improve brew formula.
Using this new update., I'm able to reconstruct/visualize old brio files (from Falaise 3.0).
Tha's looks great

@drbenmorgan : I encourage you to merge this PR to a new Falasie version to allow new 
development.

**Comment by drbenmorgan** [comment]
> I installed Falaise PR#82 based on :
> Bayeux 3.1.1
> Clang
> xdot
> We need to improve brew formula.

O.k., I'll update and test that tag of Bayeux (I have raised [Bayeux/issues/10](https://github.com/BxCppDev/Bayeux/issues/10) to help with future updates). Have you tested with GCC as well as Clang? Is `xdot` now a hard dependency? 

Could you also clarify "need to improve brew formula" please.

**Comment by lemiere** [comment]
Hi @drbenmorgan ,

Did not test using Clang.
xdot is used only for fltags output display
'improve brew formula' --> provide xdot, Clang for bayeux3.1.1.

**Comment by drbenmorgan** [comment]
> Did not test using Clang.

O.k, can do this on macOS..

> xdot is used only for fltags output display
> 'improve brew formula' --> provide xdot, Clang for bayeux3.1.1.

`xdot` and `clang-format` are already in the upstream `brew` repo (try `brew search`) so can be installed. If they are not dependencies (i.e. not needed to build), then they won't affect any of our formulae.

**Comment by drbenmorgan** [comment]
Currently blocked by Bxcppdev/Bayeux#11

**Comment by lemiere** [comment]
Hi @drbenmorgan ,

New Bayeux release 3.1.2 is available.
https://github.com/BxCppDev/Bayeux/releases/tag/3.1.2

**Comment by drbenmorgan** [comment]
Resolved on Mac by Bayeux 3.1.2, now proceeding to test that on Centos6/7 and Ubuntu 14/16. Also looks like Bayeux 3.1.2 is source but not binary compatible for Falaise 3.1.1 - that's fine for now as we'll be upgrading both in lock step.

**Comment by drbenmorgan** [comment]
Right, Bayeux 3.1.2 good on all supported platforms - now updating the docker base images for testing with the updated version.

File: source/falaise/testing/test_falaise_tags.cxx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    {
+      datatools::urn_info urnInfo;
+      std::string urn("urn:snemo:demonstrator");
+      if (dtkUrnQuery.check_urn_info(urn, falaise::tags::experiment_category())) {
+        urnInfo = dtkUrnQuery.get_urn_info(urn);
+        std::clog << "The tag '" << urn << "' is registered with category '"
+                  << urnInfo.get_category()
+                  << "'." << std::endl;
+      }
```
Comment (drbenmorgan) [inline source/falaise/testing/test_falaise_tags.cxx]: For this and the remaining blocks, what are the success/fail cases? For example, is it required that the `check_urn_info` always returns `true`? What value is expected to be returned by `get_category`? 

A basic smoke test is fine in this PR, and obviously unexpected errors are handled by the overriding `try/catch` block. In future, these tests should prefer the use of the supplied Catch infrastructure both to ease the writing of test and clarify the expected/design behaviour and conditions.

File: modules/CMakeLists.txt [hunk 3]
```diff
@@ -21,7 +21,7 @@ set(_falaise_PLUGINS
   GammaClustering
   VisuToy
   EventBrowser
-  things2root
+  #things2root
```
Comment (drbenmorgan) [inline modules/CMakeLists.txt]: Missed this - please uncomment this as the Things2ROOT module is required.

**Pull request merged**